### PR TITLE
ReducedField

### LINKED ISSUE #525: `ReducedField` type
We need a generic type that represents a reduction operation over a field that can be used both by the user to compute arbitrary reductions on the fly, and also for integration into `AbstractOperations`.

One way to accomplish this is to define a new field that looks something like

```julia
struct ReducedField{X, Y, Z, A, G, O, D, I} <: AbstractLocatedField{X, Y, Z, G}
    data :: A
    grid :: G
    reduction! :: O
    dims :: D
    input :: I
end
```

here `data` and `grid` are an array and grid similar to an ordinary `Field`, except that `data` is reduced along the tuple of dimensions in `dims`. The `input` is the 3D field / OffsetArray on which the reduction is performed. The function `reduction` performs the reduction; eg

```julia
U = ReducedField(mean, model.velocities.u, dims=(1, 2))
```

Returns an object `U` that represents the average of `model.velocities.u` over dims `(1, 2)` (x and y). We can also write wrappers / translators that allow dimensions to be referenced by name rather than number (integration with DimensionalData.jl would help in this respect). 

Computing a reduction would look something like

```julia
function compute!(reduced)
    zero_halo_regions!(parent(reduced.data), model.grid)
    reduced.reduction!(parent(reduced.data), parent(reduced.input), dims=reduced.dims)
    return nothing
end
```

Note that `mean` requires a special implementation since we have to use `sum!` followed by normalization. But this is not too onerous; we can dispatch on that scenario, and other scenarios that require special implementation not covered by the generic version above.

We can then integrate `ReducedField`s into abstract operations so we can calculate things like

```julia
u, v, w = model.velocities

U = ReducedField(mean, u, dims=(1, 2))
V = ReducedField(mean, v, dims=(1, 2))

turbulent_kinetic_energy = @at (Cell, Cell, Cell) ( (u - U)^2 + (v - V)^2 + w^2 ) / 2
```

This is a little tricky and requires the analysis of an operation tree to deduce which operations must be pre-computed. But this can be done in the constructor for `Computation` and is probably not too difficult (also, the topic of another issue).

### PR DESCRIPTION
**Pull request by glwagner**
This PR adds an `AbstractReducedField` that subtypes `AbstractField`, as well as a concrete `ReducedField` that is the analogue of `Field`. Reduced dimensions are tagged as having `Nothing` locations, and three-dimensional indexing into reduced fields is managed.

This PR also reorganizes the `Fields` submodule in light of the generalization of the concept of an `AbstractField`.

Resolves #525

### CODE REVIEW

File: src/Fields/reduced_field.jl [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+@propagate_inbounds Base.getindex(r::ARF{X, Nothing, Nothing},     i, j, k) where X = r.data[i, 1, 1]
+@propagate_inbounds Base.setindex!(r::ARF{X, Nothing, Nothing}, d, i, j, k) where X = r.data[i, 1, 1] = d
+
+@propagate_inbounds Base.getindex(r::ARF{Nothing, Y, Nothing},     i, j, k) where Y = r.data[1, j, 1]
+@propagate_inbounds Base.setindex!(r::ARF{Nothing, Y, Nothing}, d, i, j, k) where Y = r.data[1, j, 1] = d
+
+@propagate_inbounds Base.getindex(r::ARF{Nothing, Nothing, Z},     i, j, k) where Z = r.data[1, 1, k]
+@propagate_inbounds Base.setindex!(r::ARF{Nothing, Nothing, Z}, d, i, j, k) where Z = r.data[1, 1, k] = d
+
+@propagate_inbounds Base.getindex(r::ARF{Nothing, Nothing, Nothing},     i, j, k) = r.data[1, 1, 1]
```
Comment (ali-ramadhan) [inline src/Fields/reduced_field.jl:27]: 💪

File: src/Fields/reduced_field.jl [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+const DimsType = NTuple{N, Int} where N
+
+function validate_reduced_dims(dims)
+    dims = Tuple(dims)
+
+    # Check type
+    dims isa DimsType || error("Reduced dims must be an integer or tuple of integers.")
+    
+    # Check length
+    length(dims) > 3  && error("Models are 3-dimensional. Cannot average over 4+ dimensions.")
```
Comment (ali-ramadhan) [inline src/Fields/reduced_field.jl:39]:
```suggestion
    length(dims) > 3  && error("Models are 3-dimensional. Cannot reduce over 4+ dimensions.")
```

File: src/Fields/reduced_field.jl [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    return dims
+end
+
+function validate_reduced_locations(X, Y, Z, dims)
+    loc = (X, Y, Z)
+
+    # Check reduced locations
+    for i in dims
+        loc[i] != Nothing && ArgumentError("The location of reduced dimensions must be Nothing")
```
Comment (ali-ramadhan) [inline src/Fields/reduced_field.jl:52]:
```suggestion
        isnothing(loc[i]) || ArgumentError("The location of reduced dimensions must be Nothing")
```

File: src/Fields/set!.jl [hunk 4]
```diff
@@ -52,61 +52,94 @@ function set!(Φ::NamedTuple; kwargs...)
...
+const AbstractReducedCPUField =
+    AbstractReducedField{X, Y, Z, A, G} where {X, Y, Z, A<:OffsetArray{T, D, <:Array} where {T, D}, G}
 
 "Set the CPU field `u` to the array `v`."
 function set!(u::AbstractCPUField, v::Array)
-    for k in 1:u.grid.Nz, j in 1:u.grid.Ny, i in 1:u.grid.Nx
+
+    sz = size(u)
+
+    for k in 1:sz[3], j in 1:sz[2], i in 1:sz[1]
```
Comment (ali-ramadhan) [inline src/Fields/set!.jl:72]:
```suggestion
    Sx, Sy, Sz = size(u)
    for k in 1:Sz, j in 1:Sy, i in 1:Sx
```

**Review by ali-ramadhan — APPROVED** [review]
Looks pretty good!

As a side note, I think we're missing quite a few tests for `set!` that we should probably add in the future since it's a commonly used function by users.

File: src/Fields/reduced_field.jl (hunk 1 above)
Comment (glwagner) [reply src/Fields/reduced_field.jl:27]: VERY reduced field

**Pull request merged**